File: src/python/algorithms/dijkstra.py
import heapq


def dijkstra(graph, start):
    """
    Finds the shortest distance from the start node
    to every other node in the graph.
    """

    # Distance table
    distance = {}

    # Parent table
    parent = {}

    # Initialize every node
    for node in graph.nodes:
        distance[node] = float("inf")
        parent[node] = None

    # Starting point
    distance[start] = 0

    # Priority Queue
    priority_queue = []

    heapq.heappush(priority_queue, (0, start))

    while priority_queue:

        current_distance, current_node = heapq.heappop(priority_queue)

        # Ignore outdated entries
        if current_distance > distance[current_node]:
            continue

        # Explore neighbours
        for neighbour in graph.neighbors(current_node):

            edge_data = graph[current_node][neighbour]

            # Road length in meters
            weight = edge_data[0].get("length", 1)

            new_distance = current_distance + weight

            if new_distance < distance[neighbour]:

                distance[neighbour] = new_distance

                parent[neighbour] = current_node

                heapq.heappush(
                    priority_queue,
                    (new_distance, neighbour)
                )

    return distance, parent

File: src/cpp/src/Dijkstra.cpp
#include "../include/Dijkstra.h"

#include <queue>
#include <limits>
#include <functional>
#include <algorithm>

// Constructor
Dijkstra::Dijkstra(const Graph& g)
    : graph(g)
{
}

// Dijkstra Algorithm
std::unordered_map<long long, double>
Dijkstra::shortestPath(
    long long source,
    std::unordered_map<long long, long long>& parent
)
{
    std::unordered_map<long long, double> distance;

    const auto& adjacencyList = graph.getGraph();

    // Initialize distances
    for (const auto& node : adjacencyList)
    {
        distance[node.first] = std::numeric_limits<double>::infinity();
    }

    distance[source] = 0.0;
    parent[source] = -1;

    using State = std::pair<double, long long>;

    std::priority_queue<
        State,
        std::vector<State>,
        std::greater<State>
    > pq;

    pq.push({0.0, source});

    while (!pq.empty())
    {
        double currentDistance = pq.top().first;
        long long currentNode = pq.top().second;
        pq.pop();

        // Ignore outdated entries
        if (currentDistance > distance[currentNode])
        {
            continue;
        }

        // Skip nodes with no outgoing edges
        if (adjacencyList.find(currentNode) == adjacencyList.end())
        {
            continue;
        }

        // Visit neighbours
        for (const Edge& edge : adjacencyList.at(currentNode))
        {
            if (edge.blocked)
            {
                continue;
            }

            double edgeCost =
                edge.length * edge.trafficMultiplier;

            double newDistance =
                currentDistance + edgeCost;

            // Initialize destination if missing
            if (distance.find(edge.destination) == distance.end())
            {
                distance[edge.destination] =
                    std::numeric_limits<double>::infinity();
            }

            if (newDistance < distance[edge.destination])
            {
                distance[edge.destination] = newDistance;
                parent[edge.destination] = currentNode;

                pq.push({newDistance, edge.destination});
            }
        }
    }

    return distance;
}

// Reconstruct Shortest Path
std::vector<long long>
Dijkstra::reconstructPath(
    long long destination,
    const std::unordered_map<long long, long long>& parent
)
{
    std::vector<long long> path;

    long long current = destination;

    while (parent.find(current) != parent.end() &&
           parent.at(current) != -1)
    {
        path.push_back(current);
        current = parent.at(current);
    }

    path.push_back(current);

    std::reverse(path.begin(), path.end());

    return path;
}